File: src/python/graphql/strawberry/schema.py
import uuid
from datetime import datetime, timezone

import strawberry
from strawberry.types import Info

from db import check_db, get_user, get_complex_orders
from cache import check_cache, cache_get, cache_set


@strawberry.type
class Health:
    status: str
    version: str
    timestamp: str
    database: str
    cache: str


@strawberry.type
class JsonItem:
    id: int
    uuid: str
    name: str
    email: str
    created_at: str
    is_active: bool


@strawberry.type
class JsonItemsResult:
    items: list[JsonItem]
    count: int
    timestamp: str


@strawberry.type
class User:
    id: int
    email: str
    first_name: str
    last_name: str
    age: int
    created_at: str


@strawberry.type
class UserOrderStats:
    user_id: int
    user_name: str
    total_orders: int
    total_value: float
    average_order_value: float


@strawberry.type
class ComplexOrdersResult:
    period_days: int
    total_users: int
    data: list[UserOrderStats]


@strawberry.type
class CacheEntry:
    key: str
    value: str
    cached: bool
    ttl: int


@strawberry.type
class Query:
    @strawberry.field
    def health(self, info: Info) -> Health:
        db_status = check_db()
        cache_status = check_cache()
        return Health(
            status="ok",
            version="1.0.0",
            timestamp=datetime.now(timezone.utc).isoformat(),
            database=db_status,
            cache=cache_status,
        )

    @strawberry.field
    def json_items(self, limit: int = 1000) -> JsonItemsResult:
        now = datetime.now(timezone.utc).isoformat()
        items = [
            JsonItem(
                id=i,
                uuid=str(uuid.uuid4()),
                name=f"Item {i}",
                email=f"item{i}@example.com",
                created_at=now,
                is_active=i % 2 == 0,
            )
            for i in range(1, limit + 1)
        ]
        return JsonItemsResult(
            items=items,
            count=len(items),
            timestamp=now,
        )

    @strawberry.field
    def user(self, id: int) -> User | None:
        row = get_user(id)
        if row is None:
            return None
        return User(
            id=row["id"],
            email=row["email"],
            first_name=row["first_name"],
            last_name=row["last_name"],
            age=row["age"],
            created_at=row["created_at"],
        )

    @strawberry.field
    def complex_orders(self, days: int = 30) -> ComplexOrdersResult:
        data = get_complex_orders(days)
        stats = [
            UserOrderStats(
                user_id=r["user_id"],
                user_name=r["user_name"],
                total_orders=r["total_orders"],
                total_value=r["total_value"],
                average_order_value=r["average_order_value"],
            )
            for r in data
        ]
        return ComplexOrdersResult(
            period_days=days,
            total_users=len(stats),
            data=stats,
        )

    @strawberry.field
    def cache(self, key: str) -> CacheEntry:
        value, ttl = cache_get(key)
        if value is not None:
            return CacheEntry(
                key=key,
                value=value,
                cached=True,
                ttl=ttl,
            )

        generated = f"value-for-{key}"
        cache_set(key, generated, 300)
        return CacheEntry(
            key=key,
            value=generated,
            cached=False,
            ttl=300,
        )


schema = strawberry.Schema(query=Query)

File: src/dart/graphql/angel3/lib/schema.dart
// Builds the GraphQL schema using the graphql_schema2 API (the *2 fork that
// angel3_graphql depends on). The previous code imported
// `package:graphql/schema.dart`, but the resolved `graphql` package is a
// client-only distribution with no schema/executor exports; the schema types
// live in `graphql_schema2`, re-exported transitively by `angel3_graphql`.
//
// graphql_schema2 renames the classic graphql builder API:
//   GraphQLField      -> GraphQLObjectField  (positional type, not `type:`)
//   GraphQLFieldArg   -> GraphQLFieldInput   (passed via `arguments:`)
//   GraphQLNonNull(t) -> t.nonNullable()
//   GraphQLList(t)    -> listOf(t)
//   GraphQLString/Int/Float/Boolean -> graphQLString/graphQLInt/...
//   GraphQLObjectType -> objectType(name, fields: [...]) factory
// The resolver arity drops from (obj, args, ctx) to (obj, args). A resolver
// is required on every field; for leaf scalars we read the value by field
// name from the parent map.
import 'package:graphql_angel3_benchmark/canonical.dart';
import 'package:graphql_schema2/graphql_schema2.dart';

import 'db.dart';
import 'cache.dart';

/// Leaf scalar field: resolves to parent[fieldName].
GraphQLObjectField _leaf(String name, GraphQLType type) =>
    GraphQLObjectField(name, type,
        resolve: (obj, _) => (obj is Map) ? obj[name] : null);

GraphQLSchema buildSchema(DatabaseService db, CacheService cache) {
  return GraphQLSchema(
    queryType: objectType(
      'Query',
      fields: [
        GraphQLObjectField(
          'health',
          objectType('Health', fields: [
            _leaf('status', graphQLString.nonNullable()),
            _leaf('version', graphQLString.nonNullable()),
            _leaf('timestamp', graphQLString.nonNullable()),
            _leaf('database', graphQLString.nonNullable()),
            _leaf('cache', graphQLString.nonNullable()),
          ]).nonNullable(),
          resolve: (obj, args) async {
            var dbStatus = 'ok';
            var cacheStatus = 'ok';
            try {
              await db.ping();
            } catch (_) {
              dbStatus = 'error';
            }
            try {
              await cache.ping();
            } catch (_) {
              cacheStatus = 'error';
            }
            return {
              'status': 'ok',
              'version': '1.0.0',
              'timestamp': DateTime.now().toUtc().toIso8601String(),
              'database': dbStatus,
              'cache': cacheStatus,
            };
          },
        ),
        GraphQLObjectField(
          'jsonItems',
          objectType('JsonItemsResult', fields: [
            _leaf('items',
                listOf(graphQLNonNullable(objectType('JsonItem', fields: [
              _leaf('id', graphQLInt.nonNullable()),
              _leaf('uuid', graphQLString.nonNullable()),
              _leaf('name', graphQLString.nonNullable()),
              _leaf('email', graphQLString.nonNullable()),
              _leaf('createdAt', graphQLString.nonNullable()),
              _leaf('isActive', graphQLBoolean.nonNullable()),
            ])))),
            _leaf('count', graphQLInt.nonNullable()),
            _leaf('timestamp', graphQLString.nonNullable()),
          ]).nonNullable(),
          arguments: [
            GraphQLFieldInput('limit', graphQLInt, defaultValue: 1000),
          ],
          resolve: (obj, args) {
            final count = itemCount(args['limit'] as int?);
            final now = DateTime.now().toUtc().toIso8601String();
            final items = List.generate(count, (i) => {
              'id': i,
              'uuid': canonicalUuid(i),
              'name': canonicalName(i),
              'email': canonicalEmail(i),
              'createdAt': canonicalCreatedAt,
              'isActive': canonicalIsActive(i),
            });
            return {
              'items': items,
              'count': items.length,
              'timestamp': now,
            };
          },
        ),
        GraphQLObjectField(
          'user',
          objectType('User', fields: [
            _leaf('id', graphQLInt.nonNullable()),
            _leaf('email', graphQLString.nonNullable()),
            _leaf('firstName', graphQLString.nonNullable()),
            _leaf('lastName', graphQLString.nonNullable()),
            _leaf('age', graphQLInt.nonNullable()),
            _leaf('createdAt', graphQLString.nonNullable()),
          ]),
          arguments: [
            GraphQLFieldInput('id', graphQLInt.nonNullable()),
          ],
          resolve: (obj, args) async {
            return await db.getUser(args['id'] as int);
          },
        ),
        GraphQLObjectField(
          'complexOrders',
          objectType('ComplexOrdersResult', fields: [
            _leaf('periodDays', graphQLInt.nonNullable()),
            _leaf('totalUsers', graphQLInt.nonNullable()),
            _leaf(
              'data',
              listOf(graphQLNonNullable(objectType('UserOrderStats', fields: [
                _leaf('userId', graphQLInt.nonNullable()),
                _leaf('userName', graphQLString.nonNullable()),
                _leaf('totalOrders', graphQLInt.nonNullable()),
                _leaf('totalValue', graphQLFloat.nonNullable()),
                _leaf('averageOrderValue', graphQLFloat.nonNullable()),
              ]))),
            ),
          ]).nonNullable(),
          arguments: [
            GraphQLFieldInput('days', graphQLInt, defaultValue: 30),
          ],
          resolve: (obj, args) async {
            final days = args['days'] as int? ?? 30;
            final data = await db.getComplexOrders(days);
            return {
              'periodDays': days,
              'totalUsers': data.length,
              'data': data,
            };
          },
        ),
        GraphQLObjectField(
          'cache',
          objectType('CacheEntry', fields: [
            _leaf('key', graphQLString.nonNullable()),
            _leaf('value', graphQLString.nonNullable()),
            _leaf('cached', graphQLBoolean.nonNullable()),
            _leaf('ttl', graphQLInt.nonNullable()),
          ]).nonNullable(),
          arguments: [
            GraphQLFieldInput('key', graphQLString.nonNullable()),
          ],
          resolve: (obj, args) async {
            final key = args['key'] as String;
            final value = await cache.get(key);
            if (value != null) {
              final ttlVal = await cache.ttl(key);
              return {
                'key': key,
                'value': value.toString(),
                'cached': true,
                'ttl': ttlVal >= 0 ? ttlVal : 0,
              };
            }
            final generated = '{"key": "$key", "generated": true}';
            await cache.set(key, generated, 300);
            return {
              'key': key,
              'value': generated,
              'cached': false,
              'ttl': 300,
            };
          },
        ),
      ],
    ),
  );
}
